core/src/main/java/jenkins/model/run/KeepRunAction.java

```diff
@@ -0,0 +1,88 @@
+package jenkins.model.run;
+
+import hudson.Extension;
+import hudson.model.Action;
+import hudson.model.Run;
+import hudson.security.Permission;
+import java.util.Collection;
+import java.util.Set;
+import jenkins.model.TransientActionFactory;
+import jenkins.model.menu.Group;
+import jenkins.model.menu.event.Event;
+import jenkins.model.menu.event.LinkEvent;
+
+@Extension
+public class KeepRunAction extends TransientActionFactory<Run> {
+
+    @Override
+    public Class<Run> type() {
+        return Run.class;
+    }
+
+    @Override
+    public Collection<? extends Action> createFor(Run target) {
```

Comment:
I was expecting more feedback when I kept the build forever, nothing changes on the page itself unless you open the app-bar overflow and see the text has changed to allow reverting the keep forever.